db/migrate/20180528141303_fix_accounts_unique_index.rb

```diff
@@ -0,0 +1,68 @@
+class FixAccountsUniqueIndex < ActiveRecord::Migration[5.2]
+  disable_ddl_transaction!
+
+  def up
+    duplicates = Account.connection.select_all('SELECT string_agg(id::text, \',\') AS ids FROM accounts GROUP BY lower(username), lower(domain) HAVING count(*) > 1').to_hash
+
+    duplicates.each do |row|
+      deduplicate_account!(row['ids'].split(','))
+    end
+
+    remove_index :accounts, name: 'index_accounts_on_username_and_domain_lower' if index_name_exists?(:accounts, 'index_accounts_on_username_and_domain_lower')
+    safety_assured { execute 'CREATE UNIQUE INDEX CONCURRENTLY index_accounts_on_username_and_domain_lower ON accounts (lower(username), lower(domain))' }
+    remove_index :accounts, name: 'index_accounts_on_username_and_domain' if index_name_exists?(:accounts, 'index_accounts_on_username_and_domain')
+  end
+
+  def down
+    raise ActiveRecord::IrreversibleMigration
+  end
+
+  private
+
+  def deduplicate_account!(account_ids)
+    accounts          = Account.where(id: account_ids).order(updated_at: :desc)
+    reference_account = accounts.shift
+
+    accounts.each do |other_account|
```

Comment:
the each block here should be wrapped in a transaction I think?

Reply:
Is that really necessary? Even if it fails halfway through, that's still work done that won't need to be repeated on a future run.

Reply:
I'm saying each iteration of the loop should be a new transaction, so we don't lose half the progress but an error doesn't leave an account in a half-deleted state.